pkg/controllers/user/alert/statesyncer/statesyncer.go

```diff
@@ -0,0 +1,138 @@
+			if err != nil {
+				logrus.Infof("Failed to retrieve alert list from alertmanager: %v", err)
+			} else {
+				clusterAlerts, err := s.clusterAlertClient.Controller().Lister().List("", labels.NewSelector())
+				if err != nil {
+					continue
+				}
+
+				projectAlerts, err := s.projectAlertClient.Controller().Lister().List("", labels.NewSelector())
+				if err != nil {
+					continue
+				}
+
+				pAlerts := []*v3.ProjectAlert{}
+				for _, alert := range projectAlerts {
+					if controller.ObjectInCluster(s.clusterName, alert) {
+						pAlerts = append(pAlerts, alert)
+					}
+				}
+
+				for _, alert := range clusterAlerts {
+					alertID := alert.Namespace + "-" + alert.Name
+					state := s.alertManager.GetState(alertID, apiAlerts)
+					needUpdate := s.doSync(alertID, alert.Status.State, state)
+
+					if needUpdate {
+						alert.Status.State = state
+						_, err := s.clusterAlertClient.Update(alert)
+						if err != nil {
+							logrus.Errorf("Error occurred while updating alert state : %v", err)
+						}
+					}
+				}
+
+				for _, alert := range pAlerts {
+					alertID := alert.Namespace + "-" + alert.Name
+					state := s.alertManager.GetState(alertID, apiAlerts)
+					needUpdate := s.doSync(alertID, alert.Status.State, state)
+
+					if needUpdate {
+						alert.Status.State = state
+						_, err := s.projectAlertClient.Update(alert)
+						if err != nil {
+							logrus.Errorf("Error occurred while updating alert state and time: %v", err)
+						}
+					}
+				}
+			}
+		}
+	}
+
+}
+
+//The curState is the state in the CRD status,
+//The newState is the state in alert manager side
+func (s *StateSyncer) doSync(alertID, curState, newState string) (needUpdate bool) {
+	if curState == "inactive" {
+		return false
+	}
+
+	//only take ation when the state is not the same
+	if newState != curState {
+
+		//the alert is muted by user (curState == muted), but it already went away in alertmanager side (newState == active)
+		//then we need to remove the silence rule and update the state in CRD
+		if curState == "muted" && newState == "active" {
+			err := s.alertManager.RemoveSilenceRule(alertID)
+			if err != nil {
+				logrus.Errorf("Error occurred while remove silence : %v", err)
+			}
+			return true
+		}
+
+		//the alert is unmuted by user, but it is still muted in alertmanager side
+		//need to remove the silence rule, but do not have to update the CRD
+		if curState == "alerting" && newState == "muted" {
+			err := s.alertManager.RemoveSilenceRule(alertID)
+			if err != nil {
+				logrus.Errorf("Error occurred while remove silence : %v", err)
+			}
+			return false
+		}
+
+		//the alert is muted by user, but it is still alerting in alertmanager side
+		//need to add silence rule to alertmanager
+		if curState == "muted" && newState == "alerting" {
+			err := s.alertManager.AddSilenceRule(alertID)
+			if err != nil {
+				logrus.Errorf("Error occurred while remove silence : %v", err)
+			}
+			return false
```

Comment:
do we have to update CRD in this case, or returning false is ok?

Reply:
we don't have to update the CRD. This case is when user performs unmute action on the muted alert. The alert state on the CRD is changed from "muted" to "alerting" already. But the alert state on the alertmanager side is still "muted", which indicate we should remove silence rule from alertmanager. Once the rule is remove, the state on alertmanager side will become "alerting".